In [1]:
# Importar las librerías a utilizar
import pandas as pd
from datetime import datetime
import numpy as np

In [64]:
# Cargar el .csv
df_accidentes = pd.read_csv("accidentes.csv")
df_accidentes.head()

id  n_victimas       fecha  aaaa  mm  dd      hora franja_hora  \
0  2016-0001           1  2016-01-01  2016   1   1  04:00:00           4   
1  2016-0002           1  2016-01-02  2016   1   2  01:15:00           1   
2  2016-0003           1  2016-01-03  2016   1   3  07:00:00           7   
3  2016-0004           1  2016-01-10  2016   1  10  00:00:00           0   
4  2016-0005           1  2016-01-21  2016   1  21  05:20:00           5   

                                   direccion tipo_calle  ...        rol  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ    AVENIDA  ...  CONDUCTOR   
1          AV GRAL PAZ  Y AV DE LOS CORRALES   GRAL PAZ  ...  CONDUCTOR   
2                         AV ENTRE RIOS 2034    AVENIDA  ...  CONDUCTOR   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO    AVENIDA  ...  CONDUCTOR   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA     AVENIDA  ...  CONDUCTOR   

   vehiculo_victima       sexo edad             gravedad    moto auto  \
0              MOTO  MASCULINO   19  2016-01-01 00:00:00  5e-324  0.0   
1              AUTO  MASCULINO   70  2016-01-02 00:00:00  5e-324  0.0   
2              MOTO  MASCULINO   30  2016-01-03 00:00:00  5e-324  0.0   
3              MOTO  MASCULINO   18                   SD  5e-324  0.0   
4              MOTO  MASCULINO   29  2016-02-01 00:00:00  5e-324  0.0   

  transporte_publico camion ciclista  
0                0.0    0.0      0.0  
1                0.0    0.0      0.0  
2                0.0    0.0      0.0  
3                0.0    0.0      0.0  
4                0.0    0.0      0.0  

[5 rows x 31 columns]

# Fecha y Hora

Voy a conservar una columna con la fecha y una con la franja horaria.

In [65]:
# Revisar la columna de franja_horaria
df_accidentes["franja_hora"].describe()

count     28310
unique       26
top          17
freq       2032
Name: franja_hora, dtype: object

In [66]:
# Revisar la distribución de franja_horaria
df_accidentes["franja_hora"].value_counts()

franja_hora
17    2032
16    2006
14    1894
13    1857
12    1823
15    1811
18    1808
19    1562
11    1484
20    1360
9     1327
10    1300
8     1265
21    1158
0      946
22     920
7      878
23     700
6      565
1      445
5      386
2      293
4      253
3      233
sd       3
SD       1
Name: count, dtype: int64

In [67]:
# Cuento los nulos
df_accidentes["franja_hora"].isnull().sum()

5

Tengo 4 valores desconocidos y 5 nulos, voy a asignar los nulos a las 17 (5) los sd a las 16(3) y los SD a las 14 (1)

In [68]:
df_accidentes["franja_hora"].replace('sd', "16", inplace=True)
df_accidentes["franja_hora"].replace('SD', "14", inplace=True)
df_accidentes["franja_hora"].fillna("17", inplace=True)
df_accidentes["franja_hora"].describe()

count     28315
unique       24
top          17
freq       2037
Name: franja_hora, dtype: object

In [69]:
# Convertir la columna de franja_horaria a entero
df_accidentes["franja_hora"] = df_accidentes["franja_hora"].astype(np.int8)
df_accidentes["franja_hora"].describe()

count    28315.000000
mean        13.496733
std          5.634942
min          0.000000
25%         10.000000
50%         14.000000
75%         18.000000
max         23.000000
Name: franja_hora, dtype: float64

In [70]:
# Revisar la columna fecha
df_accidentes["fecha"].describe()

count                   28315
unique                   1695
top       2021-12-17 00:00:00
freq                       79
Name: fecha, dtype: object

In [71]:
# Contar los nulos
df_accidentes["fecha"].isnull().sum()

0

In [85]:
# Pasar a fecha los valores de la columna fecha
df_accidentes["fecha"] = pd.to_datetime(df_accidentes["fecha"], format='mixed')

In [86]:
# Normalizar los valores al formato que me agrade
df_accidentes["fecha"] = df_accidentes["fecha"].dt.strftime("%Y-%m-%d")

In [91]:
df_accidentes["fecha"].describe()

count          28315
unique          1446
top       2021-12-17
freq              79
Name: fecha, dtype: object

In [88]:
df_accidentes["fecha"].isnull().sum()

0

In [93]:
# Revisar la columna fecha
df_accidentes["fecha"].info

<bound method Series.info of 0        2016-01-01
1        2016-01-02
2        2016-01-03
3        2016-01-10
4        2016-01-21
            ...    
28310    2021-12-31
28311    2021-12-31
28312    2021-12-31
28313    2021-12-31
28314    2020-05-17
Name: fecha, Length: 28315, dtype: object>

Por último voy a borrar las columnas con la fecha dividida y con la hora.

In [94]:
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'aaaa', 'mm', 'dd', 'hora', 'franja_hora',
       'direccion', 'tipo_calle', 'calle', 'altura', 'cruce',
       'direccion_normalizada', 'comuna', 'geocodificacion_CABA', 'longitud',
       'latitud', 'participantes', 'victima', 'acusado', 'rol',
       'vehiculo_victima', 'sexo', 'edad', 'gravedad', 'moto', 'auto',
       'transporte_publico', 'camion', 'ciclista'],
      dtype='object')

In [95]:
# Eliminar las columnas que ya no se necesitan
columnas_eliminar = ["aaaa", "mm", "dd", "hora"]
df_accidentes = df_accidentes.drop(columnas_eliminar,axis=1)

In [96]:
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'franja_hora', 'direccion', 'tipo_calle',
       'calle', 'altura', 'cruce', 'direccion_normalizada', 'comuna',
       'geocodificacion_CABA', 'longitud', 'latitud', 'participantes',
       'victima', 'acusado', 'rol', 'vehiculo_victima', 'sexo', 'edad',
       'gravedad', 'moto', 'auto', 'transporte_publico', 'camion', 'ciclista'],
      dtype='object')

In [97]:
# Guardar el csv parcial
df_accidentes.to_csv("accidentes_parcial.csv", index=False)

# Número de Víctimas

In [99]:
# Cargar el csv
df_accidentes = pd.read_csv("accidentes_parcial.csv")

In [100]:
# Revisar la columna n_victimas
df_accidentes["n_victimas"].describe()

count    28315.000000
mean         1.413350
std          1.076277
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         16.000000
Name: n_victimas, dtype: float64

In [101]:
# Contar los valores distintos de n_victimas
df_accidentes["n_victimas"].value_counts()

n_victimas
1     21707
2      4286
3      1200
4       472
5       330
6       126
8        80
7        42
10       40
16       32
Name: count, dtype: int64

In [102]:
# Contar los valores distintos de id
df_accidentes["id"].value_counts()

id
LC-2019-0291069    16
LC-2019-0188476    16
LC-2021-0137640    10
LC-2019-0049800    10
LC-2019-0428120    10
                   ..
LC-2019-0616801     1
LC-2019-0616592     1
LC-2019-0635165     1
LC-2019-0616529     1
LC-2020-0244615     1
Name: count, Length: 24477, dtype: int64

In [104]:
# Reviso el id puntual para ver que está pasando
df_accidentes[df_accidentes["id"] == "LC-2019-0291069"]

id  n_victimas       fecha  franja_hora  \
5172  LC-2019-0291069          16  2019-05-10            7   
5173  LC-2019-0291069          16  2019-05-10            7   
5174  LC-2019-0291069          16  2019-05-10            7   
5175  LC-2019-0291069          16  2019-05-10            7   
5176  LC-2019-0291069          16  2019-05-10            7   
5177  LC-2019-0291069          16  2019-05-10            7   
5178  LC-2019-0291069          16  2019-05-10            7   
5179  LC-2019-0291069          16  2019-05-10            7   
5180  LC-2019-0291069          16  2019-05-10            7   
5181  LC-2019-0291069          16  2019-05-10            7   
5182  LC-2019-0291069          16  2019-05-10            7   
5183  LC-2019-0291069          16  2019-05-10            7   
5184  LC-2019-0291069          16  2019-05-10            7   
5185  LC-2019-0291069          16  2019-05-10            7   
5186  LC-2019-0291069          16  2019-05-10            7   
5187  LC-2019-0291069          16  2019-05-10            7   

                                              direccion tipo_calle calle  \
5172  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5173  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5174  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5175  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5176  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5177  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5178  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5179  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5180  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5181  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5182  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5183  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5184  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5185  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5186  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   
5187  Goyena, Pedro Av. y Mitre, Emilio 0, Barrio: C...         SD   NaN   

      altura cruce direccion_normalizada  ...  rol vehiculo_victima   sexo  \
5172     NaN   NaN                    SD  ...  0.0               sd  Varon   
5173     NaN   NaN                    SD  ...  0.0               sd  Varon   
5174     NaN   NaN                    SD  ...  0.0               sd  Varon   
5175     NaN   NaN                    SD  ...  0.0               sd  Varon   
5176     NaN   NaN                    SD  ...  0.0               sd  Varon   
5177     NaN   NaN                    SD  ...  0.0               sd  Varon   
5178     NaN   NaN                    SD  ...  0.0               sd  Mujer   
5179     NaN   NaN                    SD  ...  0.0               sd  Mujer   
5180     NaN   NaN                    SD  ...  0.0               sd  Mujer   
5181     NaN   NaN                    SD  ...  0.0               sd  Mujer   
5182     NaN   NaN                    SD  ...  0.0               sd  Mujer   
5183     NaN   NaN                    SD  ...  0.0               sd  Mujer   
5184     NaN   NaN                    SD  ...  0.0               sd  Mujer   
5185     NaN   NaN                    SD  ...  0.0               sd  Varon   
5186     NaN   NaN                    SD  ...  0.0               sd  Mujer   
5187     NaN   NaN                    SD  ...  0.0               sd  Varon   

     edad gravedad moto auto transporte_publico camion ciclista  
5172   SD       SD   SD   SD                  x     SD       SD  
5173   SD       SD   SD   SD                  x     SD       SD  
5174   54       SD   SD   SD                  x     SD       SD  
5175   35       SD   SD   SD                  x     SD       SD  
5176   63     

#### Como realicé la unión de los datos, hay registros duplicados en las columnas con datos globales sobre el accidente.

#### No voy a hacer cambios sobre estos datos, pero debo recordar que si quiero obtener el número de víctimas para una categoría debo sumar registros y no sumar la columna de n_victimas.

# Dirección / Ubicación Geográfica

In [2]:
# Cargar el csv
df_accidentes = pd.read_csv("accidentes_parcial.csv")

In [3]:
# Revisar que datos hay en la columna de codificación
df_accidentes["geocodificacion_CABA"].head()

0      Point (98896.78238426 93532.43437792)
1      Point (95832.05571093 95505.41641999)
2     Point (106684.29090040 99706.57687843)
3      Point (99840.65224780 94269.16534422)
4    Point (106980.32827929 100752.16915795)
Name: geocodificacion_CABA, dtype: object

In [4]:
# Revisar columnas
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'franja_hora', 'direccion', 'tipo_calle',
       'calle', 'altura', 'cruce', 'direccion_normalizada', 'comuna',
       'geocodificacion_CABA', 'longitud', 'latitud', 'participantes',
       'victima', 'acusado', 'rol', 'vehiculo_victima', 'sexo', 'edad',
       'gravedad', 'moto', 'auto', 'transporte_publico', 'camion', 'ciclista'],
      dtype='object')

De todas las columnas con información sobre el lugar del hecho, me voy a quedar con "tipo_calle", "direccion_normalizada", "comuna", "latitud" y "longitud".

In [5]:
# Distribución valores distintos
df_accidentes["tipo_calle"].value_counts()

tipo_calle
SD           12879
AVENIDA       8251
CALLE         6323
GRAL PAZ       745
AUTOPISTA      117
Name: count, dtype: int64

In [6]:
# Verificar si hay nulos
df_accidentes["tipo_calle"].isna().sum()

0

tipo_calle está OK

In [7]:
# Revisar direccion_normalizada
df_accidentes["direccion_normalizada"].describe()

count     28253
unique     9261
top          SD
freq      12643
Name: direccion_normalizada, dtype: object

In [8]:
# Revisar si hay nulos
df_accidentes["direccion_normalizada"].isna().sum()

62

In [9]:
# Rellenar los nulos
df_accidentes["direccion_normalizada"].fillna("SD", inplace=True)
df_accidentes["direccion_normalizada"].isna().sum()

0

dirección_normalizada está OK

In [10]:
# Distribución valores distintos
df_accidentes["comuna"].value_counts()

comuna
1                  2931
15                 2237
4                  2213
9                  2150
3                  2003
14                 1840
7                  1817
12                 1750
11                 1645
13                 1590
10                 1568
8                  1420
5                  1400
6                  1104
2                  1021
SD                  936
No Especificada     510
0                     2
Name: count, dtype: int64

In [117]:
# Revisar valores nulos
df_accidentes["comuna"].isna().sum()

178

Voy a juntar los valores "no Especificada", "0" y "nulos", todos en "SD" 

In [11]:
# Llenar los nulos y unificar criterio de valores desconocidos
df_accidentes["comuna"].fillna("SD", inplace=True)
df_accidentes["comuna"].replace("No Especificada", "SD", inplace=True)
df_accidentes["comuna"].replace("0", "SD", inplace=True)

In [12]:
df_accidentes["comuna"].value_counts()

comuna
1     2931
15    2237
4     2213
9     2150
3     2003
14    1840
7     1817
12    1750
11    1645
SD    1626
13    1590
10    1568
8     1420
5     1400
6     1104
2     1021
Name: count, dtype: int64

In [13]:
df_accidentes["comuna"].isna().sum()

0

Comuna está OK

In [17]:
# Revisar latitud
df_accidentes["latitud"].describe()

count     28045
unique    14362
top          SD
freq       1386
Name: latitud, dtype: object

In [16]:
# Verificar si hay nulos
df_accidentes["latitud"].isna().sum()

270

In [20]:
# Rellenar nulos con SD
df_accidentes["latitud"].fillna("SD", inplace=True)
df_accidentes["latitud"].isna().sum()

0

In [21]:
df_accidentes["latitud"].describe()

count     28315
unique    14362
top          SD
freq       1656
Name: latitud, dtype: object

In [33]:
# Revisar longitud
df_accidentes["longitud"].describe()

count     28045
unique    14415
top          SD
freq       1386
Name: longitud, dtype: object

In [34]:
# Verificar si hay nulos
df_accidentes["longitud"].isna().sum()

270

In [35]:
# Rellenar nulos con SD
df_accidentes["longitud"].fillna("SD", inplace=True)
df_accidentes["longitud"].isna().sum()

0

In [36]:
df_accidentes["longitud"].describe()

count     28315
unique    14415
top          SD
freq       1656
Name: longitud, dtype: object

Latitud y Longitud están OK

In [37]:
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'franja_hora', 'direccion', 'tipo_calle',
       'calle', 'altura', 'cruce', 'direccion_normalizada', 'comuna',
       'geocodificacion_CABA', 'longitud', 'latitud', 'participantes',
       'victima', 'acusado', 'rol', 'vehiculo_victima', 'sexo', 'edad',
       'gravedad', 'moto', 'auto', 'transporte_publico', 'camion', 'ciclista'],
      dtype='object')

In [38]:
# Eliminar las columnas que ya no se necesitan
columnas_eliminar = ["direccion", "calle", "altura", "cruce", "geocodificacion_CABA"]
df_accidentes = df_accidentes.drop(columnas_eliminar,axis=1)
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'franja_hora', 'tipo_calle',
       'direccion_normalizada', 'comuna', 'longitud', 'latitud',
       'participantes', 'victima', 'acusado', 'rol', 'vehiculo_victima',
       'sexo', 'edad', 'gravedad', 'moto', 'auto', 'transporte_publico',
       'camion', 'ciclista'],
      dtype='object')

In [41]:
# Guardar el csv parcial
df_accidentes.to_csv("accidentes_parcial.csv", index=False)

# Rol

In [67]:
# Cargar el csv
df_accidentes = pd.read_csv("accidentes_parcial.csv")

In [68]:
df_accidentes["rol"]

0        CONDUCTOR
1        CONDUCTOR
2        CONDUCTOR
3        CONDUCTOR
4        CONDUCTOR
           ...    
28310       1e-323
28311       1e-323
28312       5e-324
28313       5e-324
28314       5e-324
Name: rol, Length: 28315, dtype: object

In [69]:
df_accidentes["rol"].value_counts()

rol
0.0                     27136
CONDUCTOR                 330
5e-324                    326
PEATON                    267
PASAJERO_ACOMPAÑANTE       80
1e-323                     80
CICLISTA                   29
1.5e-323                   27
2.5e-323                   15
2e-323                     12
SD                         11
1.113730556184e-311         1
1.1137296434825e-311        1
Name: count, dtype: int64

Hay demasiados valores vacíos, y algunos de los valores se repiten con los de otras columnas. No voy a utilizar esta columna para los análisis, así que la voy a eliminar.

In [70]:
df_accidentes = df_accidentes.drop("rol",axis=1)
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'franja_hora', 'tipo_calle',
       'direccion_normalizada', 'comuna', 'longitud', 'latitud',
       'participantes', 'victima', 'acusado', 'vehiculo_victima', 'sexo',
       'edad', 'gravedad', 'moto', 'auto', 'transporte_publico', 'camion',
       'ciclista'],
      dtype='object')

In [71]:
# Guardar el csv parcial
df_accidentes.to_csv("accidentes_parcial.csv", index=False)

# Participantes / Víctima / Acusado / Vehículo víctima / Moto / Auto / Transporte Público / Camión / Ciclista

In [101]:
# Cargar el csv
df_accidentes = pd.read_csv("accidentes_parcial.csv")

En el caso del estudio de las columnas con los vehículos participantes, se fijan algunos criterios previos:
- No se va a analizar la culpabilidad de cada parte, solo se van a considerar las partes intervinientes en el siniestro vial, sin identificar responsabilidades. Por lo tanto la calificación de víctima y acusado no se mantendrá.
- La columna participantes necesita ser desagregada, porque quiero poder analizar los accidentes por vehículos intervinientes de manera individual, y si tengo por ejemplo moto-auto y moto-camión, no me sirve que sean dos datos categóricos independientes. En ambos hubo una moto involucrada y eso es lo que me interesa, lógicamente voy a guardar los datos de los dos participantes del siniestro, pero en columnas separadas.
- Las columnas de vehículo víctima y víctima tienen información casi idéntica.
- Desde uno de los excel vienen columnas diferentes para individualizar los vehículos intervinientes (moto, auto, transporte público, camión, ciclista). Como no se va a trabajar con modelos de machine learning, no considero necesario trabajar los datos de esa forma, aunque si voy a sacar datos de estas columnas, que se ven más completas que las otras.  

Como conclusión, voy a dejar dos columnas para las partes intervinientes, llamadas vehiculo_1 y vehiculo_2. Las cuales voy a rellenar con los datos del resto de las columnas de participantes, y en caso de que sean más de dos vehículos voy a sintetizarlo en los dos más relevantes. 

In [102]:
df_accidentes["participantes"]

0                      MOTO-AUTO
1                 AUTO-PASAJEROS
2                      MOTO-AUTO
3                        MOTO-SD
4                 MOTO-PASAJEROS
                  ...           
28310                      SD-SD
28311                      SD-SD
28312                      SD-SD
28313    MOTO-TRANSPORTE PUBLICO
28314                  MOTO-AUTO
Name: participantes, Length: 28315, dtype: object

In [103]:
# Dividir la columna participantes por el guion y almacenar los valores en columnas distintas
df_accidentes["participantes_1"] = df_accidentes["participantes"].str.split('-').str[0]
df_accidentes["participantes_2"] = df_accidentes["participantes"].str.split('-').str[1]
df_accidentes.head()

id  n_victimas       fecha  franja_hora tipo_calle  \
0  2016-0001           1  2016-01-01            4    AVENIDA   
1  2016-0002           1  2016-01-02            1   GRAL PAZ   
2  2016-0003           1  2016-01-03            7    AVENIDA   
3  2016-0004           1  2016-01-10            0    AVENIDA   
4  2016-0005           1  2016-01-21            5    AVENIDA   

                               direccion_normalizada comuna      longitud  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8  -58.47533969   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9  -58.50877521   
2                                ENTRE RIOS AV. 2034      1  -58.39040293   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8  -58.46503904   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1  -58.38718297   

        latitud   participantes  ...       sexo edad             gravedad  \
0  -34.68757022       MOTO-AUTO  ...  MASCULINO   19  2016-01-01 00:00:00   
1  -34.66977709  AUTO-PASAJEROS  ...  MASCULINO   70  2016-01-02 00:00:00   
2  -34.63189362       MOTO-AUTO  ...  MASCULINO   30  2016-01-03 00:00:00   
3  -34.68092974         MOTO-SD  ...  MASCULINO   18                   SD   
4  -34.62246630  MOTO-PASAJEROS  ...  MASCULINO   29  2016-02-01 00:00:00   

     moto auto transporte_publico camion ciclista participantes_1  \
0  5e-324  0.0                0.0    0.0      0.0            MOTO   
1  5e-324  0.0                0.0    0.0      0.0            AUTO   
2  5e-324  0.0                0.0    0.0      0.0            MOTO   
3  5e-324  0.0                0.0    0.0      0.0            MOTO   
4  5e-324  0.0                0.0    0.0      0.0            MOTO   

  participantes_2  
0            AUTO  
1       PASAJEROS  
2            AUTO  
3              SD  
4       PASAJEROS  

[5 rows x 23 columns]

In [104]:
# Eliminar la columna participantes
df_accidentes = df_accidentes.drop("participantes",axis=1)

In [105]:
# Revisar que valores tiene
df_accidentes["victima"].head()

0    MOTO
1    AUTO
2    MOTO
3    MOTO
4    MOTO
Name: victima, dtype: object

In [106]:
# Revisar la columna
df_accidentes["victima"].describe()

count     28315
unique       19
top          SD
freq      12831
Name: victima, dtype: object

In [107]:
# Distribución de valores únicos
df_accidentes["victima"].value_counts()

victima
SD                    12831
MOTO                   6229
AUTO                   2695
PEATON                 2422
CICLISTA               2263
TRANSPORTE PUBLICO      905
CAMIONETA               257
TAXI                    247
MOVIL                   142
MIXTO                   113
CAMION                   96
BICICLETA                29
MONOPATIN                28
OTRO                     26
UTILITARIO               16
CARGAS                    7
PASAJEROS                 5
OBJETO FIJO               2
PEATON_MOTO               2
Name: count, dtype: int64

In [108]:
# Revisar si hay nulos
df_accidentes["victima"].isna().sum()

0

In [109]:
# Revisar que valores tiene
df_accidentes["vehiculo_victima"].head()

0    MOTO
1    AUTO
2    MOTO
3    MOTO
4    MOTO
Name: vehiculo_victima, dtype: object

In [110]:
# Revisar la columna
df_accidentes["vehiculo_victima"].describe()

count     28315
unique       17
top          sd
freq      11792
Name: vehiculo_victima, dtype: object

In [111]:
# Distribución de valores únicos
df_accidentes["vehiculo_victima"].value_counts()

vehiculo_victima
sd                    11792
MOTO                   5081
SD                     4486
PEATON                 1951
CICLISTA               1759
AUTO                   1732
TRANSPORTE PUBLICO      680
TAXI                    241
UTILITARIO              231
MOVIL                   134
MIXTO                   113
CAMION                   40
BICICLETA                29
MONOPATIN                27
CARGAS                    7
OTRO                      7
PASAJEROS                 5
Name: count, dtype: int64

In [112]:
# Unificar sd y SD en la columna de vehículo_1
df_accidentes["vehiculo_victima"].replace("sd", "SD", inplace=True)

In [113]:
# Revisar si hay nulos
df_accidentes["vehiculo_victima"].isna().sum()

0

In [117]:
df_accidentes[["victima", "vehiculo_victima"]][df_accidentes["vehiculo_victima"] == "SD"].value_counts()

victima             vehiculo_victima
SD                  SD                  12831
MOTO                SD                   1149
AUTO                SD                    965
CICLISTA            SD                    504
PEATON              SD                    472
TRANSPORTE PUBLICO  SD                    225
CAMION              SD                     56
CAMIONETA           SD                     39
OTRO                SD                     19
MOVIL               SD                      8
TAXI                SD                      6
UTILITARIO          SD                      3
MONOPATIN           SD                      1
Name: count, dtype: int64

En la línea de código anterior veo que la columna victima me puede completar valores de la columna vehiculo_victima

In [118]:
# Revisar los valores de la columna de acusado
df_accidentes["acusado"]

0                      AUTO
1                 PASAJEROS
2                      AUTO
3                        SD
4                 PASAJEROS
                ...        
28310                    SD
28311                    SD
28312                    SD
28313    TRANSPORTE PUBLICO
28314                  AUTO
Name: acusado, Length: 28315, dtype: object

In [119]:
# Revisar si hay nulos 
df_accidentes["acusado"].isna().sum()

0

In [120]:
# Revisar la distribución de valores
df_accidentes["acusado"].value_counts()

acusado
SD                    17941
AUTO                   5376
TRANSPORTE PUBLICO     1188
CAMIONETA               987
MOTO                    807
TAXI                    648
OBJETO FIJO             399
CAMION                  348
PASAJEROS               178
CARGAS                  150
MOVIL                    99
CICLISTA                 90
OTRO                     31
UTILITARIO               29
MULTIPLE                 17
PEATON                   16
BICICLETA                 7
MONOPATIN                 3
TREN                      1
Name: count, dtype: int64

#### Las dos columnas de los vehículos participantes las voy a llenar inicialmente con las columnas de vehículo víctima y la columna de acusado, y luego voy a revisar si puedo completar los valores faltantes con el resto de las columnas.

In [121]:
# Renombrar las columnas
df_accidentes.rename({"vehiculo_victima": "vehiculo_1", "acusado": "vehiculo_2"}, inplace=True, axis=1)
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'franja_hora', 'tipo_calle',
       'direccion_normalizada', 'comuna', 'longitud', 'latitud', 'victima',
       'vehiculo_2', 'vehiculo_1', 'sexo', 'edad', 'gravedad', 'moto', 'auto',
       'transporte_publico', 'camion', 'ciclista', 'participantes_1',
       'participantes_2'],
      dtype='object')

-----------

In [122]:
# Guardar el csv parcial
df_accidentes.to_csv("accidentes_parcial.csv", index=False)

In [132]:
# Cargar el csv
df_accidentes = pd.read_csv("accidentes_parcial.csv")

----------------

Para preparar el relleno de las columnas vehiculo_1 y vehiculo_2, voy a llenar con nulos todos los valores SD de ambas, y el resto de las columnas las voy a dejar solamente con nulos y con el valor del vehículo cuando existe.

In [135]:
# Revisar la distribución
df_accidentes["vehiculo_1"].value_counts()

vehiculo_1
SD                    16278
MOTO                   5081
PEATON                 1951
CICLISTA               1759
AUTO                   1732
TRANSPORTE PUBLICO      680
TAXI                    241
UTILITARIO              231
MOVIL                   134
MIXTO                   113
CAMION                   40
BICICLETA                29
MONOPATIN                27
CARGAS                    7
OTRO                      7
PASAJEROS                 5
Name: count, dtype: int64

In [136]:
# Preparar los valores para los rellenos
df_accidentes["vehiculo_1"].replace('SD', np.NAN, inplace=True)

In [137]:
# Revisar la distribución
df_accidentes["vehiculo_2"].value_counts()

vehiculo_2
SD                    17941
AUTO                   5376
TRANSPORTE PUBLICO     1188
CAMIONETA               987
MOTO                    807
TAXI                    648
OBJETO FIJO             399
CAMION                  348
PASAJEROS               178
CARGAS                  150
MOVIL                    99
CICLISTA                 90
OTRO                     31
UTILITARIO               29
MULTIPLE                 17
PEATON                   16
BICICLETA                 7
MONOPATIN                 3
TREN                      1
Name: count, dtype: int64

In [138]:
# Preparar los valores para los rellenos
df_accidentes["vehiculo_2"].replace('SD', np.NAN, inplace=True)

In [139]:
# Revisar la distribución
df_accidentes["victima"].value_counts()

victima
SD                    12831
MOTO                   6229
AUTO                   2695
PEATON                 2422
CICLISTA               2263
TRANSPORTE PUBLICO      905
CAMIONETA               257
TAXI                    247
MOVIL                   142
MIXTO                   113
CAMION                   96
BICICLETA                29
MONOPATIN                28
OTRO                     26
UTILITARIO               16
CARGAS                    7
PASAJEROS                 5
OBJETO FIJO               2
PEATON_MOTO               2
Name: count, dtype: int64

In [140]:
# Preparar los valores para los rellenos
df_accidentes["victima"].replace('SD', np.NAN, inplace=True)

In [141]:
# Revisar la distribución
df_accidentes["participantes_1"].value_counts()

participantes_1
SD                    12831
MOTO                   6221
AUTO                   2694
PEATON                 2417
CICLISTA               2263
TRANSPORTE PUBLICO      905
CAMIONETA               257
TAXI                    247
MOVIL                   142
MIXTO                   113
CAMION                   96
BICICLETA                28
MONOPATIN                28
OTRO                     26
MULTIPLE                 17
UTILITARIO               16
CARGAS                    7
PASAJEROS                 5
PEATON_MOTO               2
Name: count, dtype: int64

In [142]:
# Preparar los valores para los rellenos
df_accidentes["participantes_1"].replace('SD', np.NAN, inplace=True)

In [143]:
# Revisar la distribución
df_accidentes["participantes_2"].value_counts()

participantes_2
SD                    17941
AUTO                   5374
TRANSPORTE PUBLICO     1188
CAMIONETA               987
MOTO                    807
TAXI                    648
OBJETO FIJO             401
CAMION                  348
PASAJEROS               178
CARGAS                  150
MOVIL                   102
CICLISTA                 90
UTILITARIO               29
OTRO                     28
PEATON                   16
BICICLETA                 7
MONOPATIN                 3
TREN                      1
Name: count, dtype: int64

In [145]:
# Preparar los valores para los rellenos
df_accidentes["participantes_2"].replace('SD', np.NAN, inplace=True)

Sobre las columnas siguientes, no encontré explicación de las marcas, pero infiero que "SD" es se desconoce, que "x" es para marcar en caso afirmativo, y los casos de "0" y "0.0" no sé si son positivos o negativos, pero viendo los casos de transporte público y camiones, infiero que también son casos negativos. Por lo que los únicos positivos son las "x".

In [146]:
# Revisar la distribución
df_accidentes["moto"].value_counts()

moto
SD          10153
x            9978
0            7375
5e-324        676
1e-323         38
1.5e-323        3
Name: count, dtype: int64

In [151]:
# Preparar los valores para los rellenos
df_accidentes["moto"].replace('SD', np.NAN, inplace=True)
df_accidentes["moto"].replace('0', np.NAN, inplace=True)
df_accidentes["moto"].replace('5e-324', np.NAN, inplace=True)
df_accidentes["moto"].replace('1e-323', np.NAN, inplace=True)
df_accidentes["moto"].replace('1.5e-323', np.NAN, inplace=True)
df_accidentes["moto"].replace('x', "MOTO", inplace=True)

In [147]:
# Revisar la distribución
df_accidentes["auto"].value_counts()

auto
SD     14564
x       6525
0       6417
0.0      717
Name: count, dtype: int64

In [152]:
# Preparar los valores para los rellenos
df_accidentes["auto"].replace('SD', np.NAN, inplace=True)
df_accidentes["auto"].replace('0', np.NAN, inplace=True)
df_accidentes["auto"].replace('0.0', np.NAN, inplace=True)
df_accidentes["auto"].replace('x', "AUTO", inplace=True)

In [148]:
# Revisar la distribución
df_accidentes["transporte_publico"].value_counts()

transporte_publico
SD     13539
0      10898
x       3069
0.0      717
Name: count, dtype: int64

In [153]:
# Preparar los valores para los rellenos
df_accidentes["transporte_publico"].replace('SD', np.NAN, inplace=True)
df_accidentes["transporte_publico"].replace('0', np.NAN, inplace=True)
df_accidentes["transporte_publico"].replace('0.0', np.NAN, inplace=True)
df_accidentes["transporte_publico"].replace('x', "TRANSPORTE PUBLICO", inplace=True)

In [149]:
# Revisar la distribución
df_accidentes["camion"].value_counts()

camion
SD     14755
0      11185
x       1566
0.0      717
Name: count, dtype: int64

In [163]:
# Preparar los valores para los rellenos
df_accidentes["camion"].replace('SD', np.NAN, inplace=True)
df_accidentes["camion"].replace('0', np.NAN, inplace=True)
df_accidentes["camion"].replace('0.0', np.NAN, inplace=True)
df_accidentes["camion"].replace('x', "CAMION", inplace=True)

In [150]:
# Revisar la distribución
df_accidentes["ciclista"].value_counts()

ciclista
SD     13392
0      10605
x       3509
0.0      717
Name: count, dtype: int64

In [165]:
# Preparar los valores para los rellenos
df_accidentes["ciclista"].replace('SD', np.NAN, inplace=True)
df_accidentes["ciclista"].replace('0', np.NAN, inplace=True)
df_accidentes["ciclista"].replace('0.0', np.NAN, inplace=True)
df_accidentes["ciclista"].replace('x', "CICLISTA", inplace=True)

----------------

In [167]:
# Guardar el csv parcial
df_accidentes.to_csv("accidentes_parcial.csv", index=False)

In [2]:
# Cargar el csv
df_accidentes = pd.read_csv("accidentes_parcial.csv")

-----------

A partir de los procesos que recibió el dataframe, las columnas "victima" y "participantes_1" se le van a pasar solamente a "vehiculo_1" y la columna "participantes_2" se le va a pasar a "vehiculo_2" Todas las columnas fueron preparadas para tal fin, y no se van a pasar columnas a ambos destinos para evitar que un mismo dato complete las dos columnas de vehiculo. (Ej. Si tengo un accidente entre una moto y un SD, tengo que evitar que se complete moto-moto por error.)

In [3]:
# Contar los nulos para ver la efectividad del relleno
df_accidentes["vehiculo_1"].isna().sum()

16278

In [4]:
# Rellenar usando la columna victima
df_accidentes["vehiculo_1"].fillna(df_accidentes["victima"], inplace=True)
df_accidentes["vehiculo_1"].isna().sum()


12831

In [5]:
# Rellenar usando la columna participantes_1
df_accidentes["vehiculo_1"].fillna(df_accidentes["participantes_1"], inplace=True)
df_accidentes["vehiculo_1"].isna().sum()

12831

In [6]:
# Contar los nulos para ver la efectividad del relleno
df_accidentes["vehiculo_2"].isna().sum()

17941

In [7]:
# Rellenar usando la columna participantes_2
df_accidentes["vehiculo_2"].fillna(df_accidentes["participantes_2"], inplace=True)
df_accidentes["vehiculo_2"].isna().sum()

17941

Las columnas surgidas de dividir la columna de participantes, no aportaron nuevos valores.

En el caso de las columnas de vehículos, como se marcaba con una x el vehículo involucrado, tampoco puedo saber si en el caso de una sola cruz, se trataba de 2 vehículos del mismo tipo, por lo que voy a suponer que la x solo representa un vehículo involucrado, y que si se repetía, se llenaba la columna de víctima, acusado o participantes que eran parte del mismo dataset. **Por lo tanto, voy a rellenar los valores de vehiculo_1 siempre que se pueda, y los de vehiculo_2 solo si es diferente al de vehiculo_1.**

In [10]:
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'franja_hora', 'tipo_calle',
       'direccion_normalizada', 'comuna', 'longitud', 'latitud', 'victima',
       'vehiculo_2', 'vehiculo_1', 'sexo', 'edad', 'gravedad', 'moto', 'auto',
       'transporte_publico', 'camion', 'ciclista', 'participantes_1',
       'participantes_2'],
      dtype='object')

In [8]:
# Contar los nulos para ver la efectividad del relleno
df_accidentes["vehiculo_1"].isna().sum()

12831

In [9]:
# Rellenar usando la columna moto
df_accidentes["vehiculo_1"].fillna(df_accidentes["moto"], inplace=True)
df_accidentes["vehiculo_1"].isna().sum()

9628

In [11]:
# Rellenar usando la columna auto
df_accidentes["vehiculo_1"].fillna(df_accidentes["auto"], inplace=True)
df_accidentes["vehiculo_1"].isna().sum()

9525

In [12]:
# Rellenar usando la columna transporte_publico
df_accidentes["vehiculo_1"].fillna(df_accidentes["transporte_publico"], inplace=True)
df_accidentes["vehiculo_1"].isna().sum()

8482

In [13]:
# Rellenar usando la columna camion
df_accidentes["vehiculo_1"].fillna(df_accidentes["camion"], inplace=True)
df_accidentes["vehiculo_1"].isna().sum()

8280

In [14]:
# Rellenar usando la columna ciclista
df_accidentes["vehiculo_1"].fillna(df_accidentes["ciclista"], inplace=True)
df_accidentes["vehiculo_1"].isna().sum()

7820

Arranqué con 16278 nulos y terminé con 7820, lo cual considero satisfactorio.

La función siguiente, generada con ayuda de GPT, me va a ayudar a completar las columnas del dataframe evitando duplicar el dato del vehículo implicado, según lo explicado más arriba.



In [16]:
def completar_vehiculos(df, columna_parametro):
  """
  Función para completar la columna "vehiculo_2"con el valor de la "columna_parametro" si es posible.

  Argumentos:
    df: El DataFrame con la columna "vehiculo_2" y "columna_parametro".
    columna_parametro: El nombre de la columna que se usará para completar la otra.

  Devuelve:
    El DataFrame con la columna "vehiculo_2" completadas cuando corresponda.
  """
  for i in df.index:
    if pd.isnull(df.loc[i, "vehiculo_2"]) and not pd.isnull(df.loc[i, columna_parametro]) and (df.loc[i, "vehiculo_1"] != df.loc[i, columna_parametro]):
      df.loc[i, "vehiculo_2"] = df.loc[i, columna_parametro]
  return df

In [17]:
# Contar los nulos para ver la efectividad del relleno
df_accidentes["vehiculo_2"].isna().sum()

17941

In [18]:
# Rellenar usando la columna moto
completar_vehiculos(df_accidentes, "moto")
df_accidentes["vehiculo_2"].isna().sum()

17438

In [19]:
# Rellenar usando la columna auto
completar_vehiculos(df_accidentes, "auto")
df_accidentes["vehiculo_2"].isna().sum()

17438

In [20]:
# Rellenar usando la columna transporte_publico
completar_vehiculos(df_accidentes, "transporte_publico")
df_accidentes["vehiculo_2"].isna().sum()

17006

In [21]:
# Rellenar usando la columna camion
completar_vehiculos(df_accidentes, "camion")
df_accidentes["vehiculo_2"].isna().sum()

16773

In [22]:
# Rellenar usando la columna ciclista
completar_vehiculos(df_accidentes, "ciclista")
df_accidentes["vehiculo_2"].isna().sum()

16148

Ahora voy a volver a cambiar los nulos por SD

In [25]:
df_accidentes["vehiculo_1"].fillna("SD", inplace=True)
df_accidentes["vehiculo_2"].fillna("SD", inplace=True)

In [26]:
# Revisar distribución
df_accidentes["vehiculo_1"].value_counts()

vehiculo_1
MOTO                  9433
SD                    7820
AUTO                  2800
CICLISTA              2723
PEATON                2423
TRANSPORTE PUBLICO    1948
CAMION                 298
TAXI                   247
UTILITARIO             234
MOVIL                  142
MIXTO                  113
CAMIONETA               39
BICICLETA               29
MONOPATIN               28
OTRO                    26
CARGAS                   7
PASAJEROS                5
Name: count, dtype: int64

In [27]:
# Revisar distribución
df_accidentes["vehiculo_2"].value_counts()

vehiculo_2
SD                    16148
AUTO                   5376
TRANSPORTE PUBLICO     1620
MOTO                   1310
CAMIONETA               987
CICLISTA                715
TAXI                    648
CAMION                  581
OBJETO FIJO             399
PASAJEROS               178
CARGAS                  150
MOVIL                    99
OTRO                     31
UTILITARIO               29
MULTIPLE                 17
PEATON                   16
BICICLETA                 7
MONOPATIN                 3
TREN                      1
Name: count, dtype: int64

Finalmente voy a eliminar las columnas que ya no necesito.

In [28]:
# Eliminar las columnas que ya no se necesitan
columnas_eliminar = ["victima", "participantes_1", "participantes_2", "moto", "auto", "transporte_publico", "camion", "ciclista"]
df_accidentes = df_accidentes.drop(columnas_eliminar,axis=1)
df_accidentes.columns

Index(['id', 'n_victimas', 'fecha', 'franja_hora', 'tipo_calle',
       'direccion_normalizada', 'comuna', 'longitud', 'latitud', 'vehiculo_2',
       'vehiculo_1', 'sexo', 'edad', 'gravedad'],
      dtype='object')

#### Pude eliminar más de la mitad de los nulos del primer vehículo implicado en el accidente, y solo unos pocos del segundo vehículo implicado.

In [29]:
# Guardar el csv parcial
df_accidentes.to_csv("accidentes_parcial.csv", index=False)

In [ ]:
# Cargar el csv
df_accidentes = pd.read_csv("accidentes_parcial.csv")

# Sexo